helpers/derhelpers/ed25519.go

```diff
@@ -0,0 +1,111 @@
+package derhelpers
+
+import (
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"errors"
+
+	"golang.org/x/crypto/ed25519"
+)
+
+var errEd25519WrongID = errors.New("incorrect object identifier")
+
+// ed25519OID is the OID for the Ed25519 signature scheme: see
+// https://datatracker.ietf.org/doc/draft-ietf-curdle-pkix-04.
+var ed25519OID = asn1.ObjectIdentifier{1, 3, 101, 112}
+
+// subjectPublicKeyInfo reflects the ASN.1 object defined in the X.509 standard.
+//
+// This is defined in crypto/x509 as "publicKeyInfo".
+type subjectPublicKeyInfo struct {
+	Algorithm pkix.AlgorithmIdentifier
+	PublicKey asn1.BitString
+}
+
+// MarshalEd25519PublicKey creates a DER-encoded SubjectPublicKeyInfo for an
+// ed25519 public key, as defined in
+// https://tools.ietf.org/html/draft-ietf-curdle-pkix-04. This is analagous to
+// MarshalPKIXPublicKey in crypto/x509, which doesn't currently support Ed25519.
+func MarshalEd25519PublicKey(pk ed25519.PublicKey) ([]byte, error) {
+	spki := subjectPublicKeyInfo{
+		Algorithm: pkix.AlgorithmIdentifier{
+			Algorithm: ed25519OID,
+		},
+		PublicKey: asn1.BitString{
+			BitLength: len(pk) * 8,
+			Bytes:     pk,
+		},
+	}
+
+	return asn1.Marshal(spki)
+}
+
+// ParseEd25519PublicKey returns the Ed25519 public key encoded by the input.
+func ParseEd25519PublicKey(der []byte) (ed25519.PublicKey, error) {
+	var spki subjectPublicKeyInfo
+	if rest, err := asn1.Unmarshal(der, &spki); err != nil {
+		return nil, err
+	} else if len(rest) > 0 {
+		return nil, errors.New("SubjectPublicKeyInfo too long")
```

Comment:
cfssl generally uses its own internal error codes for this. You can see [this](https://github.com/cloudflare/cfssl/blob/master/helpers/derhelpers/derhelpers.go#L27) as an example in this package.

Reply:
I'm not sure if cferrs is appropriate here. This function is analogous to:
https://golang.org/src/crypto/x509/x509.go?s=1352:1421#L45

This is is meant to surface low-level details about the encoding. I suppose I could use `cferr.Wrap()` to wrap these low-level errors, but I'd need to add a `PublicKeyError` category. (`CertificateError isn't really appropriate.)

Reply:
I'm not sure either; the other function returns a `cferr`, but we can probably get by without it here. @cbroglie do you have any thoughts on this?

Reply:
There's a mix of `errors` and `cferr` in helpers.go, I'm ok with this